// @forge/kvs/out/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForgeKvsAPIError = exports.ForgeKvsError = void 0;
const version_1 = require("./version");
class ForgeKvsError extends Error {
    packageVersion;
    constructor(message, packageVersion = version_1.PACKAGE_VERSION) {
        super(message);
        this.name = 'ForgeKvsError';
        this.packageVersion = packageVersion;
    }
}
exports.ForgeKvsError = ForgeKvsError;
class ForgeKvsAPIError extends ForgeKvsError {
    responseDetails;
    code;
    message;
    context;
    constructor(responseDetails, forgeError) {
        super(forgeError.message);
        const { status, statusText, traceId, httpMethod, httpPath, responseBodyLength } = responseDetails;
        this.responseDetails = { status, statusText, traceId, httpMethod, httpPath, responseBodyLength };
        const { code, message, context, ...bodyData } = forgeError;
        this.code = code;
        this.message = message;
        this.context = { ...context, ...bodyData };
    }
}
exports.ForgeKvsAPIError = ForgeKvsAPIError;
