// @forge/kvs/out/kvs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KvsImpl = void 0;
const entity_1 = require("./entity");
const query_1 = require("./query");
const transaction_api_1 = require("./transaction-api");
class KvsImpl {
    storageApi;
    constructor(storageApi) {
        this.storageApi = storageApi;
    }
    get(key, options) {
        return this.storageApi.get({ key, options });
    }
    set(key, value, options) {
        return this.storageApi.set({ key, value, options });
    }
    batchSet(items) {
        return this.storageApi.batchSet(items);
    }
    batchDelete(items) {
        return this.storageApi.batchDelete(items);
    }
    batchGet(items) {
        return this.storageApi.batchGet(items);
    }
    delete(key) {
        return this.storageApi.delete({ key });
    }
    query(options) {
        return new query_1.KvsQueryBuilder(this.storageApi, options);
    }
    getSecret(key, options) {
        return this.storageApi.getSecret({ key, options });
    }
    setSecret(key, value, options) {
        return this.storageApi.setSecret({ key, value, options });
    }
    deleteSecret(key) {
        return this.storageApi.deleteSecret({ key });
    }
    entity(entityName) {
        return new entity_1.EntityImpl(entityName, this.storageApi);
    }
    transact() {
        return new transaction_api_1.TransactionBuilderImpl(this.storageApi);
    }
}
exports.KvsImpl = KvsImpl;
